### ISSUE TITLE
Delay JS execution time (LazyLoad Script)

### ISSUE THREAD

**Issue by GeekPress** [wp-media/wp-rocket#2899]
## What are we proposing to do? 

With this feature WP Rocket will delay the execution of JavaScript files. This will essentially be a LazyLoad for JavaScript.

**Note**: We once planned a similar feature, but only for third party scripts (#[<u>2581</u>](https://github.com/wp-media/wp-rocket/issues/2581)). With this feature we are extending this to both internal and external JS files. 

## Why are we doing this? 

Not all scripts are relevant for the initial page render and delaying them can improve the loading time. For example, a script useful for an “Exit Popup” is not relevant during initial page load and can be delayed. This can have huge performance gains as evident from the Benchmark tests. 

As a side effect, delaying the execution of JavaScript will help improve PageSpeed score and speed test results. 

<p></p>

**Note**: We decided not to go ahead with #2581 because we thought of a better way to do it. However, that approach will take some development time and we are hoping the current approach will serve as an intermediate solution for our customers. 

## Benchmark Tests

Jonathan has been using this feature on our own website for a while with impressive results. We also did detailed benchmarking for #[<u>2581</u>](https://github.com/wp-media/wp-rocket/issues/2581) and are available in the issue. 

## User expectations and experience. How will we do this feature?

We are going to keep it simple and lean for our first iteration.   

- We will do it with a simple checkbox to enable the option. 

- The whole feature will be opt-in. Meaning that only the scripts that the users have selected will be delayed. 

- Enabling the option will reveal a text area where users can enter the scripts that they wish to delay. 

- We will pre-populate this textarea with some scripts that we have identified as safe to delay. Example: Facebook Social Plugins.

- Users can add their own scripts (both internal and external) to the list. They can also specify a keyword from the inline JS to be delayed. 

- If any of the scripts we include by default causes issues, users can remove them from the list as well. 

- A link « Restore default settings » will restore the default list if the user clicks on it. This will provide users an easy way to restore the default list. 

<p></p>

**Updating the default list as we discover more scripts that can be safely delayed**

<p></p>

- For now, let’s not do an automatic update for existing users. We can update our documentation and maintain a public list. Users can copy from our public list when needed. After they set-up the page initially, it wouldn’t make sense to modify the list. 

- Once we have more feedback and also see how we can handle this feature in other ways, we can work on an update strategy. Let’s keep this lean for now. 

- The updated list can be included for new users.

- This updated list can also be included if the user deletes all the scripts from the list and save settings. 

<p></p>

**UI proposal**

![](https://lh6.googleusercontent.com/ebVtOsmUG0vSTUZtp3KA8jakMxIemUtMPd3zfGt77PGB6_KCjww5RaoLCdM-4fYGJol9ISmmt4lq1OA7Cyc1XEvgr8gZhfGP8V71mb5ISOxRam3sciMyLAktPIYS7H-D8cyXkLCa)

<p></p>

- Option name: **Delay JavaScript execution**

- Option description: **Improves performance by delaying the loading of JavaScript files until user interaction (e.g. scroll, click). More info**

- Option location: **WP Rocket > File Optimization tab > after “Load JavaScript deferred”**

- Text area title: **Scripts to delay**

- Text are description: **Specify keywords that can identify inline or JavaScript files to be delayed (one per line).**

<p></p>

**New users:**

- This option will be turned on by default on new installations.

<p></p>

**Both existing and new users:** 

- Enabling “Safe Mode” will disable the feature.

- This option will be applied on uncached pages too.

## Pre-selected scripts

Please refer this doc for suggestions: <https://www.notion.so/wpmedia/95155dc2f51f47fd94f3c942096a02d7?v=970b622c689843f9b2aa5235bf394420>

<p></p>

We will be including this as the default list for all users. 

## Technical Specs

The following is meant to be a starting point for the dev team found during research. 

<p></p>

**A good starting point is Flying Scripts WordPress plugin -** [<u>**https://wordpress.org/plugins/flying-scripts/**</u>](https://wordpress.org/plugins/flying-scripts/)

<p></p>

This plugin does the following conversion for scripts and replaces them during runtime. Similar to what we do for LazyLoad.

```
<script async defer crossorigin="anonymous" src="https://connect.facebook.net/en_GB/sdk.js#xfbml=1&version=v6.0"></script>
```

to

```
<script async defer crossorigin="anonymous" data-type="lazy" data-src="https://connect.facebook.net/en_GB/sdk.js#xfbml=1&version=v6.0"></script>
```

<p></p>

**And for inline JS:**

```
<script>(function(d, s, id) {
var js, fjs = d.getElementsByTagName(s)[0];
if (d.getElementById(id)) return;
js = d.createElement(s); js.id = id;
js.src = "https://connect.facebook.net/en_US/sdk.js#xfbml=1&version=v2.5";
fjs.parentNode.insertBefore(js, fjs);
}(document, 'script', 'facebook-jssdk'));</script>

```

To

```
<script data-type="lazy" data-src="data:text/javascript;base64,KGZ1bmN0aW9uKGQsIHMsIGlkKSB7DQogICAgdmFyIGpzLCBmanMgPSBkLmdldEVsZW1lbnRzQnlUYWdOYW1lKHMpWzBdOw0KICAgIGlmIChkLmdldEVsZW1lbnRCeUlkKGlkKSkgcmV0dXJuOw0KICAgIGpzID0gZC5jcmVhdGVFbGVtZW50KHMpOyBqcy5pZCA9IGlkOw0KICAgIGpzLnNyYyA9ICJodHRwczovL2Nvbm5lY3QuZmFjZWJvb2submV0L2VuX1VTL3Nkay5qcyN4ZmJtbD0xJnZlcnNpb249djIuNSI7DQogICAgZmpzLnBhcmVudE5vZGUuaW5zZXJ0QmVmb3JlKGpzLCBmanMpOw0KICB9KGRvY3VtZW50LCAnc2NyaXB0JywgJ2ZhY2Vib29rLWpzc2RrJykpOw=="></script>
```

<p></p>

**Note**: Flying Scripts adds a delay (default 5 seconds) or waits for user interaction (mouse movement, scroll etc) with the website, whichever is first. Let’s default to an “infinite” delay, which means the scripts will only be loaded on first user interaction.   

( Another method of handling inline JS is described here: [<u>https://stackoverflow.com/questions/8946715/lazy-loading-javascript-and-inline-javascript</u>](https://stackoverflow.com/questions/8946715/lazy-loading-javascript-and-inline-javascript) )

<p></p>

Swift performance Lite has this feature and it handles it a bit differently. The third party JS is included and delayed from within in the combined JS file. 

- To test this, go to Swift Performance Lite > Settings tab

- Switch to advanced view - [<u>https://jmp.sh/4X4denu</u>](https://jmp.sh/4X4denu)

- Under optimization > scripts. Enable “Merge Scripts” - [<u>https://jmp.sh/1s5JpfK</u>](https://jmp.sh/1s5JpfK)

- Add a keyword from the script to LazyLoad - [<u>https://jmp.sh/tXmasBj</u>](https://jmp.sh/tXmasBj)

## Risks

We are not expecting any major risk since the entire feature is enabled manually by the user. 

<p></p>

The list of scripts that we include by default are tested and shouldn’t cause any issue. But if that is the case, users can easily remove it by simply deleting it from the UI

## Acceptance Criteria

- On activation of the setting, the text area will be pre-populated with the scripts that we have identified as safe to delay.

- New scripts can be added and existing scripts can be removed from the text area. 

- If a keyword matches with a JavaScript file, this file won't be minified / concatenated and deferred. If a keyword matches an inline JS, this script won't be inserted into the concatenated file if the « Combine JavaScript files » option is activated.

- These scripts will not be loaded when the page loads until a user interaction. 

- On user interaction, the script will be loaded. 

- Restore the default scripts list when a user clicks on « Restore default settings ». 

<p></p>

**Testing notes:** 

- Cross browser testing is recommended for this to make sure we don’t break things on different browsers. 

## Documentation and Translations

- A new document to be created explaining the feature. 

- The documentation will list the current exclusions and we can maintain this list with new scripts that are safe to be delayed.

**Comment by GeekPress** [comment]
Link to feature: https://wp-media.productboard.com/feature-board/planning/features/4982356

**Comment by Tabrisrp** [comment]
How is this expected to work with our combine JS, and defer JS options?

**Comment by GeekPress** [comment]
@Tabrisrp In the acceptance criteria we have conditions for that:

> If a keyword matches with a JavaScript file, this file won't be minified / concatenated and deferred. If a keyword matches an inline JS, this script won't be inserted into the concatenated file if the « Combine JavaScript files » option is activated.

**Comment by hellofromtonya** [comment]
@Tabrisrp @engahmeds3ed I suggest that we use just one data attribute instead of 2 and make ours unique. Why? To avoid conflicts with 3rd party code.

A suggestion is to use `data-rocketlazyloadscript` and then set the script's `src` value as the `data` attribute value.

**Comment by hellofromtonya** [comment]
For the inline JavaScript code, we are encoding the JS code (not the `<script>` tag) into base64. The script that runs in the browser will not decode it from base64 back to raw JS code. 

Why?

1. Risk of encoding issues between server-side encoding to base64 and the decoding in the browser back to a string. If there's an encoding difference in a browser, it will break the JS code.
2. Converting the JS code to base64 is also an optimization as it results in less bytes.

@Tabrisrp brought up a point about technical savvy customers noticing that the JS code is not restored in the browser. (It stays in base64 format.) He's right. That could cause confusion and support tickets.

I think the base64 is an additional (albeit likely small) optimization. If we share that in our documentation and change log, then it's part of the feature.

@GeekPress and @arunbasillal what do you think? Are you okay with us leaving it in base64 format?

**Comment by GeekPress** [comment]
@hellofromtonya I have no idea of the impact. If we know that it safer to not revert the inline JS from base64 to normal, let's keep them into a base64.

**Comment by hellofromtonya** [comment]
@GeekPress Good. Thanks. Yes, we'll leave it in base64 then and ensure our test cases are complete.

**Comment by hellofromtonya** [comment]
The script for this feature is dependent upon PR https://github.com/wp-media/wp-rocket/pull/2984. We'll need to remove the duplicate code and make the delay js script dependent upon and use the browser checker.

I'll take this task.

**Comment by arunbasillal** [comment]
I updated the Option Description to the following:

> Option description: Improves performance by delaying the loading of JavaScript files until user interaction (e.g. scroll, click). More info

Based on this feedback - https://wp-media.slack.com/archives/CFSCRLVGB/p1597348523001300

**Comment by amiremami** [comment]
Is this same as the feature in NITROCDN?

**Comment by kasket091** [comment]
Can the dev do something like after all script have been delayed then if no interaction for 1 sec or custom time on it and will load all the script since it looks like broken.
